### PR TITLE
feat: v4 to v5 migration for DNS records

### PR DESCRIPTION
**Pull request by tamas-jozsa**
Add v4 to v5 migration for DNS records

  Summary

  Implements comprehensive migration support for DNS records from Terraform Provider
  v4 to v5, handling the resource rename from cloudflare_record to
  cloudflare_dns_record along with all associated schema changes.

  Changes

  Migration Implementation

  - Resource rename: cloudflare_record → cloudflare_dns_record
  - Field migrations:
    - Renamed value → content for record values
    - Converted all numeric fields from TypeInt to Float64Attribute
    - Added required ttl field with default value of 1
    - Removed deprecated allow_overwrite field

  Complex Transformations

  - CAA record handling: Special logic for transforming data field structure
  - Priority hoisting: Extracted priority from nested data blocks for MX/SRV/URI
  records
  - Type conversions: Dynamic handling for fields that can be string or numeric (e.g.,
   CAA flags)
  - Data structure changes: Array to object transformations for complex record types

  State Migration

  - Comprehensive state transformation supporting both v4 JSON state format and v5
  requirements
  - Graceful handling of missing/null fields
  - Automatic timestamp generation for v5 compatibility

  Test Coverage

  - Coverage: 93.1% of statements
  - Test suite: 43 comprehensive test cases covering:
    - Basic resource renaming
    - Field transformations (renames, removals, additions)
    - Complex data structure migrations
    - Edge cases (null values, missing fields, empty arrays)
    - Multiple resource configurations
    - State file transformations

  Testing

  go test -v -race ./internal/resources/dns_record/...
  # PASS - All 43 tests passing
  # Coverage: 93.1%

  Impact

  This migration ensures seamless upgrade path for users transitioning from Terraform
  Provider v4 to v5, preserving all DNS record configurations and state while adapting
   to the new schema requirements.

### CODE REVIEW

File: internal/init/init.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package init
+
+// Import all resource packages to trigger their init() functions
```
Comment (mgirouard) [inline internal/init/init.go]: I thought we had agreed not to use `init()` anymore.
See the feedback in https://github.com/cloudflare/tf-migrate/pull/1

**Review by mgirouard — CHANGES_REQUESTED** [review]

File: internal/resources/dns_record/register.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"github.com/cloudflare/tf-migrate/internal"
+	"github.com/cloudflare/tf-migrate/internal/transform"
+)
+
+// init registers the DNS record migrators
+func init() {
+	// Register v4 to v5 migrator with explicit version
+	internal.Register("cloudflare_dns_record", "v4", "v5", func() transform.ResourceTransformer {
```
Comment (mgirouard) [inline internal/resources/dns_record/register.go]: Could these be simplified to
```go
	internal.Register("cloudflare_dns_record", "v4", "v5", NewV4ToV5Migrator)
```

**Comment by mgirouard** [comment]
Nit: on the commit; this isn't really a `chore` it's a `feat`

File: internal/resources/dns_record/v4_to_v5.go [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+}
+
+// NewV4ToV5Migrator creates a new DNS record migrator
+func NewV4ToV5Migrator() transform.ResourceTransformer {
+	migrator := &V4ToV5Migrator{
+		BaseResourceTransformer: resources.NewBaseResourceTransformer("cloudflare_dns_record"),
+	}
+
+	// Set custom handlers
+	migrator.BaseResourceTransformer.CanHandleFunc = migrator.canHandle
```
Comment (mgirouard) [inline internal/resources/dns_record/v4_to_v5.go]: This feels a bit clunky no? Could we simply reference the `migrator` instance in the `BaseResourceTransformer` instead of having to wire it up every time?

**Comment by vaishakdinesh** [comment]
`nit`: More of a strong nit, can we not name the file `v4_to_v5.go`? I'd rather have the exposed functionality which states the intent.

File: internal/resources/base.go [hunk 4]
```diff
@@ -8,29 +8,55 @@ import (
 	"github.com/cloudflare/tf-migrate/internal/transform"
 )
 
+// ResourceMigrator is a marker interface for concrete migrators.
+// Migrators can implement any combination of the following optional interfaces:
+// - ResourceHandleChecker: to customize CanHandle behavior
+// - ResourceConfigTransformer: to transform HCL configurations
+// - ResourceStateTransformer: to transform Terraform state
+// - ResourcePreprocessor: to preprocess content before parsing
+type ResourceMigrator interface{}
```
Comment (vaishakdinesh) [inline internal/resources/base.go]: Little unclear why we need this empty interface and type assertions below?

File: internal/resources/base.go [hunk 5]
```diff
@@ -8,29 +8,55 @@ import (
...
-func NewBaseResourceTransformer(resourceType string) *BaseResourceTransformer {
+func NewBaseResourceTransformer(resourceType string, migrator ResourceMigrator) *BaseResourceTransformer {
+	return &BaseResourceTransformer{
+		ResourceType: resourceType,
+		Migrator:     migrator,
+	}
+}
+
+// NewLegacyBaseResourceTransformer creates a base transformer using the old function pointer pattern
+// Deprecated: Use NewBaseResourceTransformer with a ResourceMigrator interface instead
```
Comment (vaishakdinesh) [inline internal/resources/base.go]: Why are we deprecating?

File: internal/resources/dns_record/registry.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"github.com/cloudflare/tf-migrate/internal"
+	"github.com/cloudflare/tf-migrate/internal/transform"
+)
+
+// RegisterMigrations registers all DNS record migrations with the internal registry.
+// This function is called from internal/resources/register_all.go during initialization.
+func RegisterMigrations() {
+	internal.Register("cloudflare_dns_record", "4", "5", NewV4ToV5Migrator)
+	internal.Register("cloudflare_record", "4", "5", NewV4ToV5Migrator)
```
Comment (vaishakdinesh) [inline internal/resources/dns_record/registry.go]: Do we need both?

File: internal/resources/dns_record/v4_to_v5_test.go [hunk 7]
```diff
@@ -0,0 +1,1120 @@
...
+								"modified_on": "2024-01-01T00:00:00Z"
+							}
+						}]
+					}]
+				}`,
+			},
+		},
+	}
+
+	testhelpers.RunResourceTestSuite(t, suite, NewV4ToV5Migrator)
```
Comment (vaishakdinesh) [inline internal/resources/dns_record/v4_to_v5_test.go]: Not a huge fan of helper methods which run entire test suites. This kinda prohibits custom assertions, feels a little a difficult to understand what the test is doing with out actually going into the method and checking.

All for test helpers in general which expose functionality for standard assertions, wiring up dependencies etc. 

thoughts @mgirouard @musa-cf

File: internal/registry/registry.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+import (
+	"github.com/cloudflare/tf-migrate/internal/resources/dns_record"
+)
+
+// RegisterAllMigrations registers all resource migrations with the internal registry.
+// This function should be called once during initialization to set up all available migrations.
+// Each resource package has its own RegisterMigrations function that defines how to register
+// its specific migrations.
+func RegisterAllMigrations() {
```
Comment (vaishakdinesh) [inline internal/registry/registry.go:11]: Can this just not live in the main file?

File: internal/registry/registry.go [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+	dns_record.RegisterMigrations()
```
Comment (vaishakdinesh) [inline internal/registry/registry.go]: probably better to have it `RegisterV4ToV5Migrations`

File: internal/resources/base.go [hunk 10]
```diff
@@ -40,30 +66,55 @@ func (t *BaseResourceTransformer) GetResourceType() string {
 }
 
 // TransformConfig handles configuration file transformations
-// returns the original content if no ConfigTransformer is set
 func (t *BaseResourceTransformer) TransformConfig(ctx *transform.Context, block *hclwrite.Block) (*transform.TransformResult, error) {
+	// Try interface method first
+	if t.Migrator != nil {
```
Comment (vaishakdinesh) [inline internal/resources/base.go]: Why are doing this? When will this be nil and having to have to fall back on the legacy?
Also what is the notion of legacy at this point?

File: internal/registry/registry.go (hunk 8 above)
Comment (vaishakdinesh) [reply internal/registry/registry.go:11]: Even as simple as a slice of migrators and iterating through them to invoke them

File: internal/resources/base.go [hunk 11]
```diff
@@ -8,29 +8,35 @@ import (
...
 type BaseResourceTransformer struct {
-	ResourceType  string
-	CanHandleFunc func(resourceType string) bool
-
-	// Transformation functions (optional - can be set directly)
-	ConfigTransformer func(*transform.Context, *hclwrite.Block) (*transform.TransformResult, error)
-	StateTransformer  func(*transform.Context, gjson.Result, string) (string, error)
-	Preprocessor      func(string) string
+	ResourceType string
+	Migrator     ResourceMigrator // Reference to the concrete migrator
```
Comment (vaishakdinesh) [inline internal/resources/base.go]: Why are we embedding this?

File: internal/resources/base.go (hunk 10 above)
Comment (tamas-jozsa) [reply internal/resources/base.go]: good catch!

File: internal/resources/dns_record/registry.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
+package dns_record
+
+import (
+	"github.com/cloudflare/tf-migrate/internal"
+	"github.com/cloudflare/tf-migrate/internal/transform"
+)
+
+// RegisterMigrations registers all DNS record migrations with the internal registry.
+// This function is called from internal/resources/register_all.go during initialization.
+func RegisterMigrations() {
```
Comment (vaishakdinesh) [inline internal/resources/dns_record/registry.go]: I think this should be version specific. `RegisterV4ToV5Migrations`

File: internal/registry/registry.go (hunk 9 above)
Comment (tamas-jozsa) [reply internal/registry/registry.go]: RegisterMigrations can have multiple migration paths. Like 4->5, 5->6. We should keep dns_record.RegisterMigrations()

File: internal/registry/registry.go (hunk 8 above)
Comment (tamas-jozsa) [reply internal/registry/registry.go:11]: I'm not sure. This function will have 300+ lines eventually. I'd rather keep it separate from main.

File: internal/resources/dns_record/registry.go (hunk 12 above)
Comment (tamas-jozsa) [reply internal/resources/dns_record/registry.go]: we'll have other migration paths defined here in the future

File: internal/resources/dns_record/v4_to_v5_test.go (hunk 7 above)
Comment (tamas-jozsa) [reply internal/resources/dns_record/v4_to_v5_test.go]: I'd rather not repeat the function over hundreds of resources. I think in our case, we can expect that the test structures are going to be very similar if not identical. That is, we have a **source config & state** we apply migrator and we have an **expected config & state**.

Custom assertions can still be added if the situation requires though.

File: internal/resources/base.go (hunk 11 above)
Comment (tamas-jozsa) [reply internal/resources/base.go]: Sent in an update. The goal here is that the resource migrators will implement this interface: https://github.com/cloudflare/tf-migrate/pull/4/files#diff-4eb094bff4fd6f1f9172348979bc45b1b890539eee8b26c5770643cdf9ed3dd7R13

And this base migrator will call them.

That legacy thing was just some noise in the process.

File: cmd/tf-migrate/main.go [hunk 13]
```diff
@@ -330,22 +335,23 @@ func findTerraformFiles(dir string) ([]string, error) {
...
 		return fmt.Errorf("unsupported migration path: %s", versionPath)
 	}
-
 	return nil
 }
 
 func getProviders(resources ...string) transform.Provider {
 	getFunc := func(resourceType string, source string, target string) transform.ResourceTransformer {
-		return internal.GetMigrator(resourceType, source, target)
+		sourceVersion, _ := strconv.Atoi(source)
```
Comment (mgirouard) [inline cmd/tf-migrate/main.go]: We're potentially throwing out errors here. Does this get easier and safer if we change the types to `uint`s?

Comment (tamas-jozsa) [reply cmd/tf-migrate/main.go]: It just feels cleaner to work with numbers than strings. Version numbers will always be integers, so it just make sense to use integers.

Also, we don't worry about errors here, because they are not possible, since we validate version paths before: https://github.com/cloudflare/tf-migrate/blob/0b19943ffe3ca80b0e5de96439e9ba7f2b173401/cmd/tf-migrate/main.go#L337

**Review by vaishakdinesh — APPROVED** [review]

**Review by mgirouard — APPROVED** [review]
